In [58]:
import pandas as pd

In [59]:
subscribers = pd.read_csv("subscribers.csv")
titles = pd.read_csv("titles.csv")
watch_history = pd.read_csv("watch_history.csv")
ratings = pd.read_csv("ratings.csv")
reviews = pd.read_csv("reviews.csv")
watchlist = pd.read_csv("watchlist.csv")

In [60]:
#For each table, run a basic profile: row count, column count, and data types.

In [61]:
subscribers.info()
titles.info()
watch_history.info()
ratings.info()
reviews.info()
watchlist.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   subscriber_id      15000 non-null  object 
 1   signup_date        15000 non-null  object 
 2   country            15000 non-null  object 
 3   region             15000 non-null  object 
 4   age                15000 non-null  int64  
 5   gender             15000 non-null  object 
 6   plan_type          15000 non-null  object 
 7   monthly_price_usd  15000 non-null  float64
 8   household_size     15000 non-null  int64  
 9   primary_device     15000 non-null  object 
 10  payment_method     15000 non-null  object 
 11  tenure_months      15000 non-null  int64  
 12  is_active          15000 non-null  bool   
 13  churn_date         3801 non-null   object 
dtypes: bool(1), float64(1), int64(3), object(9)
memory usage: 1.5+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90

In [62]:
#Check for missing values in every column — report which columns have nulls and what % of rows are affected

In [63]:
print("Subscribers")
print(subscribers.isnull().sum())

print("\nTitles")
print(titles.isnull().sum())

print("\nWatch_History")
print(watch_history.isnull().sum())

print("\nRatings")
print(ratings.isnull().sum())

print("\nReviews")
print(reviews.isnull().sum())

print("\nWatchlist")
print(watchlist.isnull().sum())

Subscribers
subscriber_id            0
signup_date              0
country                  0
region                   0
age                      0
gender                   0
plan_type                0
monthly_price_usd        0
household_size           0
primary_device           0
payment_method           0
tenure_months            0
is_active                0
churn_date           11199
dtype: int64

Titles
title_id                   0
title_name                 0
type                       0
primary_genre              0
country                    0
language                   0
release_year               0
date_added                 0
maturity_rating            0
seasons                    0
content_duration_min       0
is_original                0
license_type               0
director                   0
cast                       0
quality_score              0
popularity_score           0
license_cost_usd           0
license_expiry          1966
total_watch_hours          0
total_pla

In [64]:
print("Churn Date Missing % :", round(subscribers["churn_date"].isnull().mean()*100,2), "%")
print("\nLicense Expiry Missing % :", round(titles["license_expiry"].isnull().mean()*100,2), "%")

Churn Date Missing % : 74.66 %

License Expiry Missing % : 21.84 %


## Observation

- Most columns across the StreamFlix datasets do not contain any missing values.

- The `churn_date` column in the `subscribers` table contains **74.66% missing values**. These are expected because active subscribers do not have a churn date. Therefore, these missing values will be retained and will not be treated as a data-quality error.

- The `license_expiry` column in the `titles` table contains **21.84% missing values**. These missing values are expected for StreamFlix Originals, as Originals do not require a licence expiry date. Therefore, these values will also be retained.

- No other missing-value issues were identified during the initial missing-value check.

- Therefore, no missing-value imputation or deletion is required at this stage. The expected business-related missing values will be retained.


In [65]:
#Check for duplicate rows — especially duplicate watch_id in watch_history.csv and subscriber_id in subscribers.csv.

In [66]:
print("Subscribers")
print(subscribers.duplicated().sum())

print("\nTitles")
print(titles.duplicated().sum())

print("\nWatch_History")
print(watch_history.duplicated().sum())

print("\nRatings")
print(ratings.duplicated().sum())

print("\nReviews")
print(reviews.duplicated().sum())

print("\nWatchlist")
print(watchlist.duplicated().sum())

Subscribers
0

Titles
0

Watch_History
0

Ratings
0

Reviews
0

Watchlist
0


In [67]:
print("Duplicate Watch IDs :", watch_history["watch_id"].duplicated().sum())
print("Duplicate Subscriber IDs :", subscribers["subscriber_id"].duplicated().sum())

Duplicate Watch IDs : 0
Duplicate Subscriber IDs : 0


## Observation

- No duplicate rows were found in any of the six StreamFlix tables.

- The `subscribers` table contains no duplicate subscriber records, and the `titles` table contains no duplicate title records.

- The `watch_history` table contains no duplicate viewing-session records based on the duplicate-row check. In particular, no duplicate `watch_id` records were identified.

- The `ratings`, `reviews`, and `watchlist` tables also contain no duplicate rows.

- Therefore, no duplicate-row removal is required at this stage.

In [68]:
#Validate data types — watch_date and signup_date should be dates, not strings; durations and prices should be numeric.

In [69]:
print("Subscribers")
print(subscribers.dtypes)

print("\nTitles")
print(titles.dtypes)

print("\nWatch History")
print(watch_history.dtypes)

print("\nRatings")
print(ratings.dtypes)

print("\nReviews")
print(reviews.dtypes)

print("\nWatchlist")
print(watchlist.dtypes)

Subscribers
subscriber_id         object
signup_date           object
country               object
region                object
age                    int64
gender                object
plan_type             object
monthly_price_usd    float64
household_size         int64
primary_device        object
payment_method        object
tenure_months          int64
is_active               bool
churn_date            object
dtype: object

Titles
title_id                 object
title_name               object
type                     object
primary_genre            object
country                  object
language                 object
release_year              int64
date_added               object
maturity_rating          object
seasons                   int64
content_duration_min      int64
is_original                bool
license_type             object
director                 object
cast                     object
quality_score           float64
popularity_score        float64
license_cost_usd

In [70]:
# Subscribers
subscribers["signup_date"] = pd.to_datetime(subscribers["signup_date"], format="%d-%m-%Y")
subscribers["churn_date"] = pd.to_datetime(subscribers["churn_date"], format="%d-%m-%Y")

# Titles
titles["date_added"] = pd.to_datetime(titles["date_added"], format="%d-%m-%Y")
titles["license_expiry"] = pd.to_datetime(titles["license_expiry"], format="%d-%m-%Y")

# Watch History
watch_history["watch_date"] = pd.to_datetime(watch_history["watch_date"], format="%d-%m-%Y")

# Ratings
ratings["rating_date"] = pd.to_datetime(ratings["rating_date"], format="%d-%m-%Y")

# Reviews
reviews["review_date"] = pd.to_datetime(reviews["review_date"], format="%d-%m-%Y")

# Watchlist
watchlist["added_date"] = pd.to_datetime(watchlist["added_date"], format="%d-%m-%Y")

## Observation

- The date columns were initially stored as `object` and were successfully converted to `datetime64[ns]` using the actual date format present in the CSV files (`DD-MM-YYYY`).

- The numeric columns, including subscriber age, monthly subscription price, content duration, watch duration, completion percentage, ratings, and other quantitative fields, are already stored as appropriate numeric data types (`int64` or `float64`).

- No additional numeric type conversion was required.

- Therefore, the required date and numeric data-type validation has been completed successfully.

In [71]:
#Check for outliers in watch_duration_min — are there sessions longer than the content itself? Flag them.

In [72]:
duration_outliers = watch_history[watch_history["watch_duration_min"] > watch_history["content_duration_min"]]

print("Number of sessions with watch duration greater than content duration:", len(duration_outliers))

Number of sessions with watch duration greater than content duration: 0


## Observation

- No sessions were found where `watch_duration_min` exceeded `content_duration_min`.

- A total of **0 sessions (0.00%)** were flagged as duration anomalies.

- This indicates that the recorded viewing duration is within the expected content duration limit for all viewing sessions.

- Therefore, no outlier records were identified from this business-rule check, and no correction is required.

In [73]:
#Verify referential integrity — do all subscriber_id and title_id values in watch_history.csv exist in their parent tables?

In [74]:
# 1. Watch History → Subscribers

invalid_watch_subscriber_ids = watch_history[~watch_history["subscriber_id"].isin(subscribers["subscriber_id"])]

print("Watch History → Subscribers")
print("Invalid Subscriber IDs :", len(invalid_watch_subscriber_ids))

if len(invalid_watch_subscriber_ids) == 0:
    print("All subscriber_ids in watch_history.csv exist in subscribers.csv.")
else:
    print("Some subscriber_ids in watch_history.csv do not exist in subscribers.csv.")

# 2. Watch History → Titles

invalid_watch_title_ids = watch_history[~watch_history["title_id"].isin(titles["title_id"])]

print("\nWatch History → Titles")
print("Invalid Title IDs :", len(invalid_watch_title_ids))

if len(invalid_watch_title_ids) == 0:
    print("All title_ids in watch_history.csv exist in titles.csv.")
else:
    print("Some title_ids in watch_history.csv do not exist in titles.csv.")

# 3. Ratings → Subscribers

invalid_rating_subscriber_ids = ratings[~ratings["subscriber_id"].isin(subscribers["subscriber_id"])]

print("\nRatings → Subscribers")
print("Invalid Subscriber IDs :", len(invalid_rating_subscriber_ids))

if len(invalid_rating_subscriber_ids) == 0:
    print("All subscriber_ids in ratings.csv exist in subscribers.csv.")
else:
    print("Some subscriber_ids in ratings.csv do not exist in subscribers.csv.")

# 4. Ratings → Titles

invalid_rating_title_ids = ratings[~ratings["title_id"].isin(titles["title_id"])]

print("\nRatings → Titles")
print("Invalid Title IDs :", len(invalid_rating_title_ids))

if len(invalid_rating_title_ids) == 0:
    print("All title_ids in ratings.csv exist in titles.csv.")
else:
    print("Some title_ids in ratings.csv do not exist in titles.csv.")

# 5. Reviews → Subscribers

invalid_review_subscriber_ids = reviews[~reviews["subscriber_id"].isin(subscribers["subscriber_id"])]

print("\nReviews → Subscribers")
print("Invalid Subscriber IDs :", len(invalid_review_subscriber_ids))

if len(invalid_review_subscriber_ids) == 0:
    print("All subscriber_ids in reviews.csv exist in subscribers.csv.")
else:
    print("Some subscriber_ids in reviews.csv do not exist in subscribers.csv.")


# 6. Reviews → Titles

invalid_review_title_ids = reviews[~reviews["title_id"].isin(titles["title_id"])]

print("\nReviews → Titles")
print("Invalid Title IDs :", len(invalid_review_title_ids))

if len(invalid_review_title_ids) == 0:
    print("All title_ids in reviews.csv exist in titles.csv.")
else:
    print("Some title_ids in reviews.csv do not exist in titles.csv.")

# 7. Watchlist → Subscribers

invalid_watchlist_subscriber_ids = watchlist[~watchlist["subscriber_id"].isin(subscribers["subscriber_id"])]

print("\nWatchlist → Subscribers")
print("Invalid Subscriber IDs :", len(invalid_watchlist_subscriber_ids))

if len(invalid_watchlist_subscriber_ids) == 0:
    print("All subscriber_ids in watchlist.csv exist in subscribers.csv.")
else:
    print("Some subscriber_ids in watchlist.csv do not exist in subscribers.csv.")


# 8. Watchlist → Titles

invalid_watchlist_title_ids = watchlist[~watchlist["title_id"].isin(titles["title_id"])]

print("\nWatchlist → Titles")
print("Invalid Title IDs :", len(invalid_watchlist_title_ids))

if len(invalid_watchlist_title_ids) == 0:
    print("All title_ids in watchlist.csv exist in titles.csv.")
else:
    print("Some title_ids in watchlist.csv do not exist in titles.csv.")

Watch History → Subscribers
Invalid Subscriber IDs : 0
All subscriber_ids in watch_history.csv exist in subscribers.csv.

Watch History → Titles
Invalid Title IDs : 0
All title_ids in watch_history.csv exist in titles.csv.

Ratings → Subscribers
Invalid Subscriber IDs : 0
All subscriber_ids in ratings.csv exist in subscribers.csv.

Ratings → Titles
Invalid Title IDs : 0
All title_ids in ratings.csv exist in titles.csv.

Reviews → Subscribers
Invalid Subscriber IDs : 0
All subscriber_ids in reviews.csv exist in subscribers.csv.

Reviews → Titles
Invalid Title IDs : 0
All title_ids in reviews.csv exist in titles.csv.

Watchlist → Subscribers
Invalid Subscriber IDs : 0
All subscriber_ids in watchlist.csv exist in subscribers.csv.

Watchlist → Titles
Invalid Title IDs : 0
All title_ids in watchlist.csv exist in titles.csv.


## Observation

- Referential integrity was verified across all child tables and their respective parent tables.

- All `subscriber_id` values in `watch_history`, `ratings`, `reviews`, and `watchlist` were found in the `subscribers` table. No invalid or orphan subscriber IDs were identified.

- All `title_id` values in `watch_history`, `ratings`, `reviews`, and `watchlist` were found in the `titles` table. No invalid or orphan title IDs were identified.

- Therefore, all foreign-key relationships are valid across the StreamFlix datasets, and no referential-integrity issues were identified.

- No records required removal or correction due to invalid foreign-key references.

##Specific checks to run:

In [75]:
#subscribers.csv: For churned subscribers, confirm churn_date is after signup_date. Flag any row where it is not.

In [76]:
invalid_churn_dates = subscribers[subscribers["churn_date"] <= subscribers["signup_date"]]
print("Invalid Churn Dates:", len(invalid_churn_dates))

Invalid Churn Dates: 0


## Observation

- **0 invalid records** were found where `churn_date` was earlier than or equal to `signup_date`.

- Therefore, all identified churned subscribers have a valid churn date occurring after their signup date.

In [77]:
#subscribers.csv: Confirm active subscribers (is_active = True) have a blank churn_date. Flag mismatches.

In [78]:
active_churn_mismatch = subscribers[(subscribers["is_active"] == True) & (subscribers["churn_date"].notna())]

print("Active Subscribers with Churn Date:", len(active_churn_mismatch))

Active Subscribers with Churn Date: 0


## Observation

- The number of active subscribers was compared with the number of subscribers having a blank `churn_date`.

- Both counts were found to be equal.

- Therefore, all active subscribers (`is_active = True`) have a blank `churn_date`, and no mismatch was identified.

In [79]:
#watch_history.csv: Confirm completion_pct roughly equals watch_duration_min ÷ content_duration_min × 100.

In [80]:
avg_completion_pct = round(watch_history["completion_pct"].mean(),2)

print("Avg Completion Pct:", avg_completion_pct,"%")

Avg Completion Pct: 65.31 %


In [81]:
overall_duration = round(watch_history["watch_duration_min"].sum()/watch_history["content_duration_min"].sum()*100,2)

print("Avg Duration:", overall_duration,"%")

Avg Duration: 65.23 %


## Observation

- The average recorded `completion_pct` is **65.31%**, while the calculated overall completion percentage based on `watch_duration_min` and `content_duration_min` is **65.23%**.

- The difference is only **0.08 percentage points**, indicating that the recorded completion percentage is approximately consistent with the calculated value.

- Therefore, the `completion_pct` values are considered valid for this quality check, and no issue was identified.

In [82]:
#reviews.csv: Confirm sentiment only contains Positive / Neutral / Negative. Flag anything else

In [83]:
valid_sentiments = reviews[(reviews["sentiment"]== "Positive") | (reviews["sentiment"]== "Negative") | (reviews["sentiment"]== "Neutral")]
print("Valid Sentiments:", len(valid_sentiments))
print("Total Reviews:", len(reviews))

Valid Sentiments: 110000
Total Reviews: 110000


## Observation

- The `sentiment` column was checked against the allowed values: `Positive`, `Neutral`, and `Negative`.

- The number of records containing valid sentiment values was equal to the total number of reviews.

- Therefore, all review records contain valid sentiment categories, and no invalid sentiment values were identified.

In [84]:
#watchlist.csv: Does every watchlist entry link to a real subscriber and a real title?

In [85]:
# 1. Check Watchlist → Subscribers

invalid_watchlist_subscriber_ids = watchlist[~watchlist["subscriber_id"].isin(subscribers["subscriber_id"])]

print("Watchlist → Subscribers")
print("Invalid Subscriber Links :", len(invalid_watchlist_subscriber_ids))

if len(invalid_watchlist_subscriber_ids) == 0:
    print("Every watchlist entry links to a real subscriber.")
else:
    print("Some watchlist entries do not link to a real subscriber.")


# 2. Check Watchlist → Titles

invalid_watchlist_title_ids = watchlist[~watchlist["title_id"].isin(titles["title_id"])]

print("\nWatchlist → Titles")
print("Invalid Title Links :", len(invalid_watchlist_title_ids))

if len(invalid_watchlist_title_ids) == 0:
    print("Every watchlist entry links to a real title.")
else:
    print("Some watchlist entries do not link to a real title.")

Watchlist → Subscribers
Invalid Subscriber Links : 0
Every watchlist entry links to a real subscriber.

Watchlist → Titles
Invalid Title Links : 0
Every watchlist entry links to a real title.


## Observation

- Every watchlist entry was checked to confirm that its `subscriber_id` exists in the `subscribers` table.

- No invalid subscriber links were identified.

- Every watchlist entry was also checked to confirm that its `title_id` exists in the `titles` table.

- No invalid title links were identified.

- Therefore, every watchlist entry is linked to a valid subscriber and a valid title.

In [86]:
#subscribers.csv: Are there subscribers with very short tenure but marked active? That is fine — just note it.

In [100]:
active_subscribers = subscribers[subscribers["is_active"] == True]
very_short_tenure_subscription = active_subscribers["tenure_months"].min()

print("Minimum Tenure of Active Subscribers:",very_short_tenure_subscription, "months")

Minimum Tenure of Active Subscribers: 1 months


In [102]:
short_tenure_active = subscribers[(subscribers["is_active"] == True) & (subscribers["tenure_months"] == 1)]

print("Active Subscribers with Tenure 1 Months:",len(short_tenure_active))

Active Subscribers with Tenure 1 Months: 321


## Observation

- The minimum tenure among active subscribers is **1 month**.

- Active subscribers with the minimum tenure were identified and counted.

- Short-tenure active subscribers are considered acceptable according to the project requirements and are therefore not treated as data-quality issues.

- No correction or removal is required for these records.

# Phase 1 — Data Quality Report

## StreamFlix Data Quality Summary

The StreamFlix dataset consists of six interconnected tables: `subscribers`, `titles`, `watch_history`, `ratings`, `reviews`, and `watchlist`. The data was reviewed using Pandas to assess completeness, duplicates, data types, outliers, referential integrity, and the specific business rules defined for the project.

### 1. Missing Values

The missing-value check showed that only two columns contain missing values:

- `subscribers.churn_date` contains missing values for active subscribers. These are expected because active subscribers do not have a churn date.
- `titles.license_expiry` contains missing values for StreamFlix Originals, for which a licence expiry date is not applicable.

No other missing-value issues were identified. Therefore, no missing-value imputation or deletion was required.

### 2. Duplicate Records

Duplicate-row checks were performed across all six tables. No duplicate rows were identified, including the important records in `watch_history` and `subscribers`. Therefore, no duplicate records required removal.

### 3. Data Type Validation

The date columns were converted from `object` to `datetime64[ns]` using the date format present in the CSV files. The numeric columns were already stored using appropriate `int64` or `float64` data types, so no additional numeric conversion was required.

### 4. Outlier Check

`watch_duration_min` was checked against `content_duration_min` to identify viewing sessions where the recorded watch duration exceeded the full content duration. No such sessions were identified, so no duration anomalies required correction.

### 5. Referential Integrity

Foreign-key relationships were validated between all event tables and their parent tables. All `subscriber_id` values in `watch_history`, `ratings`, `reviews`, and `watchlist` were found in the `subscribers` table. Similarly, all `title_id` values were found in the `titles` table. No invalid or orphan foreign-key references were identified.

### 6. Business Rule Validation

Several business rules were checked:

- All churned subscribers had a `churn_date` occurring after their `signup_date`, with no invalid records identified.
- Active subscribers were confirmed to have a blank `churn_date`.
- The recorded `completion_pct` was broadly consistent with the relationship between watch duration and content duration.
- All review sentiment values were within the allowed categories: `Positive`, `Neutral`, and `Negative`.
- Watchlist subscriber and title relationships were valid.
- Active subscribers with short tenure were identified. The minimum tenure among active subscribers was 1 month. These records are acceptable according to the project requirements and were not treated as data-quality issues.

## Overall Conclusion

The StreamFlix datasets are generally clean and suitable for further analysis. The identified missing values are expected business-related blanks rather than data-quality errors. No duplicate records, invalid foreign-key relationships, viewing-duration anomalies, or invalid sentiment categories were identified. The required date fields were converted to appropriate datetime types, while the existing numeric fields were already correctly typed. Therefore, no major data-cleaning corrections are required before proceeding to Phase 2 — Exploratory Data Analysis (EDA).